More anchors + new features (WIP): :hide-header:, :nested: complete, :post-process:

TomerGodinger opened the pull request:
This adds the following features:
1. **`:hide-header:` option**: omits the title and description from the output.
2. **`:nested: complete` option**: produces a fuller listing with title, description, usage, short summary and then full nested details. Feel free to change the name to whatever you want if you choose to keep it. =)
3. **`:post-process: path.to.module:func_name` option**: invokes the specified function in the specified module on the list of generated nodes for each command. The function should get the command used and the list of nodes that were generated.

This also adds quite a few new anchors for cross-referencing - namely one for each `.. program::` directive, and for options and environment variables, an anchor is added per option name. For example, for a command group `grp`, command `cmd`, an option that has the flags `-f` and `--flag` and an environment variable `env` that provides a default for it, the following anchors will be created:
* `grp-cmd-f`
* `grp-cmd-flag`
* `grp-cmd-f-env`
* `grp-cmd-flag-env`

This has a caveat which I haven't dealt with, which is collisions.
Having an option with the flag names `-f` and `--f` would cause the same anchor to be created twice. Similarly a command called `cmd` with the flag `--some-flag` and a command called `cmd-some` with the flag `--flag` would both create the anchor `cmd-some-flag`. It may be a good idea to change the format so it's injective.
This actually applies to existing anchors as well, e.g. group `grp` + command `some-cmd` + option `opt` would yield the same option anchor as group `grp-some` + command `cmd` + option `opt`, if I'm not mistaken.
A decision not to support this would be valid, of course, but it's something that begs consideration at the least.

As mentioned in the title, this is not a finished, polished product.
While it does seem to work in general, I haven't tested it thoroughly, and since it adds new anchors, it currently breaks nearly all the existing tests, as they don't expect to get them.

stephenfin commented on sphinx_click/ext.py:
```diff
@@ -223,14 +223,20 @@ def _format_envvar(
-        yield ''
+        # Add an anchor for each form of parameter name
+        # For click.option('--flag', '-f', ...) it'll create anchors for "flag" and "f"
+        param_names = list(set([param_name.lstrip('-') for param_name in param.opts] + [param.name]))
```
Sets are unordered so you'll lose ordering here. It'll also make output inconsistent breaking reproducible builds. I suspect we either should use lists entirely or sort the set (producing a list)

stephenfin replied:
Also, you don't want to do this here. This code is for environment variables (see function name)

TomerGodinger replied:
I used a set in order to avoid repeating the same anchor twice.
I didn't think about the ordering - I figured it wouldn't matter for the actual output, but it does matter for tests, so I agree that it should be sorted at the least.
I don't quite get what you mean about not wanting to do this here though - this generates anchors for the environment variables. Why not put it in the code for environment variables?

stephenfin commented on sphinx_click/ext.py:
```diff
@@ -491,8 +495,6 @@ def _generate_nodes(
         """
         ctx = click.Context(command, info_name=name, parent=parent)
-        if nested == NESTED_COMPLETE:
-            print("Nested with COMPLETE")
```
Can you avoid adding this in the prior commit, please

TomerGodinger replied:
Oh damn, I'm really sorry, I was sure I'd removed all that stuff...
Will be removed, of course.

stephenfin commented on sphinx_click/ext.py:
```diff
@@ -162,14 +172,17 @@ def _format_options(ctx: click.Context) -> ty.Generator[str, None, None]:
+    command_name = ctx.command_path.replace(' ', '-')
     for param in params:
-        for line in _format_option(param):
+        for line in _format_option(param, command_name):
```
Instead of passing through this can you either (a) generate the label here, like we do for envvars in `_format_envvars` or (b) pass through the context (`ctx`) as the first parameter? We should also have a `_format_command_name(ctx)` helper to ensure we do the formatting consistently

TomerGodinger replied:
I do think that generating the label inside `_format_option()` makes more sense as it's part of the generated content for each specific option, which is what that function should be in charge of.
Passing the context is just as good though; I'll do that instead.
I agree about the extra function. I'll add that as well.

stephenfin replied:
> I do think that generating the label inside `_format_option()` makes more sense as it's part of the generated content for each specific option, which is what that function should be in charge of.

:+1: If we're doing this, can we do the same for `_format_envvar` (i.e. move the logic there from `_format_envvars`)

> Passing the context is just as good though; I'll do that instead. I agree about the extra function. I'll add that as well.

:+1:

stephenfin commented on sphinx_click/ext.py:
```diff
@@ -162,14 +172,17 @@ def _format_options(ctx: click.Context) -> ty.Generator[str, None, None]:
 
 
-def _format_argument(arg: click.Argument) -> ty.Generator[str, None, None]:
+def _format_argument(arg: click.Argument, command_name: str) -> ty.Generator[str, None, None]:
```
ditto

stephenfin commented on sphinx_click/ext.py:
```diff
@@ -428,74 +497,126 @@ def _generate_nodes(
             else:
                 commands = _filter_commands(ctx, commands)
-                for command in commands:
+                for subcommand in commands:
```
Although we use the term "subcommands" above I'm not sure it's correct. For example, if listing commands in a group, these are really commands, not subcommands. Can we keep the existing name?

TomerGodinger replied:
If you'd like to use another name, that's fine, but `command` is the name of the parameter you get in the `_generate_nodes()` function, of which I added a use at the end of the function to invoke the post-processor.
This loop was overriding the contents of said parameter, so that instead of the post-processor being invoked with the original command, it was instead invoked with the last value that was put in that variable during this loop.
So we can either:
1. Rename this loop variable.
2. Rename the original parameter.
3. Make a copy of the original parameter in another variable before performing the loop.
Personally I figured renaming the loop variable makes the most sense so I went with that.
Please tell me which option you prefer and I'll change it accordingly.

stephenfin replied:
Fair point. Let's stick with `subcommand` and just add a comment indicating that these are really commands if `command` is a group?

stephenfin commented on sphinx_click/ext.py:
```diff
@@ -515,7 +636,7 @@ def run(self) -> ty.Iterable[nodes.section]:
             ]
 
-        return self._generate_nodes(prog_name, command, None, nested, commands)
+        return self._generate_nodes(prog_name, command, None, nested, commands, hide_header=hide_header)
```
Don't use the named kwarg here. It's not necessary

TomerGodinger replied:
Yes, I added that because of something I later removed and forgot to change it back.
I'll change it back.

stephenfin replied:
Ack

stephenfin reviewed (changes requested):
A good start. Got a good few suggestions inline. Tests would really help a lot here. Also, I see a couple of cases where you add things in the first patch and remove them again in the second or third one. Can you rebase and clean these up before force pushing updates?

TomerGodinger commented:
Thanks for all the suggestions.
I'll make the changes when I can and submit another PR (or can I update this one instead?).
Sorry for the mess. This is the first PR I've made and I'm not well-versed in how it works.

TomerGodinger commented:
> A good start. Got a few suggestions inline. Tests would really help a lot here. Also, I see a couple of cases where you add things in the first patch and remove them again in the second or third one. Can you rebase and clean these up before force pushing updates?

Can you please tell me which cases of adding and removing the same things you're referring to?
The only such thing I've found was that debug print, which needed to disappear altogether.
I don't see anything else of the sort.

TomerGodinger commented:
@stephenfin Just a friendly reminder that I'm waiting for your replies before I update this PR.

stephenfin reviewed (changes requested):
Sorry for the delayed response. Let me know if anything is unclear and thanks for the reminder.

stephenfin commented:
> I've pushed an update based on everything we've discussed. Also fixed the tests and added another one for the `complete` nesting. Tell me what you think when you can please.

Oh, I missed this reply. Will look later today or tomorrow 🙈

stephenfin commented:
@TomerGodinger I reworked a few of the commits and fixed broken type checks but the general content should be the same. Could you take a look over and see if you're happy with it? Could you update the docs just to reflect the new changes? Once that's done, I think we're good to go.

PS: You can probably drop `WIP` from the PR title now too :smile:

TomerGodinger commented:
Hey, sorry for the delay.
I haven't had time to do this, but I'll try to get on it as soon as I can.

lorenzo-cavazzi commented on sphinx_click/ext.py:
```diff
@@ -438,74 +519,134 @@ def _generate_nodes(
 
-        return [section]
+        final_nodes = []
+        if hide_header:
+            final_nodes = subcommand_nodes
+
+            if nested == NESTED_NONE or nested == NESTED_SHORT:
+                section = nodes.paragraph()
+                self.state.nested_parse(result, 0, section)
+                final_nodes.insert(0, section)
+
+        else:
+            # Title
+
+            section = nodes.section(
+                '',
+                nodes.title(text=name),
+                ids=[nodes.make_id(ctx.command_path)],
+                names=[nodes.fully_normalize_name(ctx.command_path)],
+            )
+
+            sphinx_nodes.nested_parse_with_titles(self.state, result, section)
+
+            for node in subcommand_nodes:
+                section.append(node)
+            final_nodes = [section]
```
I feel the behavior here is inconsistent depending on `:nested:` `short` / `full`.
With `short`, the headers are hidden as expected (left is without `hide-header`)

![Screenshot_20220805_152116](https://user-images.githubusercontent.com/43481553/183086728-73f442d4-3ad5-45c1-b744-26adcf678acd.png)

With `full`, the headers are still hidden, but the main command also disappears:
in the following example, the missing command is
> renku project [OPTIONS] COMMAND [ARGS]...

![Screenshot_20220805_152211](https://user-images.githubusercontent.com/43481553/183087148-68270165-4140-412c-ba62-9d93b97ea818.png)

I thought this could be circumvented by combining  `:nested: none` and `:hide-header:`, but the result is slightly different since part of the header is still shown:
in the following example, the additional text is
> Project commands.

![Screenshot_20220805_152657](https://user-images.githubusercontent.com/43481553/183087539-af27f087-b6f8-4728-b617-c2e07013b48d.png)

lorenzo-cavazzi reviewed (commented):
Sorry for interfering; I was looking for something similar to `hide-header` and I found your PR. These are all great additions!

I believe there is a small inconsistency when using `hide-header`, see the comment inline.

sravel commented:
hello I want to use the option :hide-header:
can you fix and valid merge request?

sravel commented:
will this request be included?

stephenfin commented:
@sravel Someone needs to pick this up again as I assume @TomerGodinger no longer needs this.

stephenfin commented:
@TomerGodinger No issues. There's no panic with this so I'm happy to wait.

stephenfin commented:
Closing this for now. If you're able to get back to this, please feel free to reopen.

TomerGodinger commented:
OK, I realized that I'm probably never going to get to dive back into this, so I did enough to at least get it to a reasonable state.
I rebased on the current `master` branch, fixed conflicts, updated the tests, fixed the `short`/`full` nesting discrepancy that @lorenzo-cavazzi pointed out, and added documentation for the new things.

@stephenfin I don't know how to reopen this PR, so I'd appreciate your guidance there. I actually don't know how to get it to sync the new changes, either. The branch I made in my fork already has them, but I don't see them here.
I can open a new PR if preferable.

I'm sure more can and should be added here (perhaps more tests), but I don't know if and when I'll get to that.
It's a bit of a problem for me to actually run this on my system right now.
If someone wants to add some, I can at least take a look at them.

I'll try to pay more attention to this until we're done with it, but I have a pretty high load nowadays so it's problematic. I had a bit of downtime today so I wanted to make this workable again at least. Only took me two years. :sweat_smile:

stephenfin commented:
> @stephenfin I don't know how to reopen this PR, so I'd appreciate your guidance there. I actually don't know how to get it to sync the new changes, either. The branch I made in my fork already has them, but I don't see them here.
I can open a new PR if preferable.

I think you probably want to open a new PR at this point. You can link it back here. I'll do my best to review it sooner rather than later, so we can finally get this closed off!

TomerGodinger commented:
^ Done.
